docs/spec/connection-handshake/interfaces.md

```diff
@@ -0,0 +1,168 @@
+3. using new abstraction + parameter flipping at relayer
+
+
+### 1. Original
+
+From [ICS 003](https://github.com/cosmos/ics/tree/35e911fa8a9a465a03c73b72b2a3399c15e9ce16/spec/ics-003-connection-semantics):
+
+```typescript
+function connOpenTry(
+  desiredIdentifier: Identifier,
+  counterpartyConnectionIdentifier: Identifier,
+  counterpartyPrefix: CommitmentPrefix,
+  counterpartyClientIdentifier: Identifier,
+  clientIdentifier: Identifier,
+  counterpartyVersions: string[],
+  proofInit: CommitmentProof,
+  proofConsensus: CommitmentProof,
+  proofHeight: uint64,
+  consensusHeight: uint64) {
+    abortTransactionUnless(validateConnectionIdentifier(desiredIdentifier))
+    abortTransactionUnless(consensusHeight <= getCurrentHeight())
+    expectedConsensusState = getConsensusState(consensusHeight)
+    expected = ConnectionEnd{INIT, desiredIdentifier, getCommitmentPrefix(), counterpartyClientIdentifier,
+                             clientIdentifier, counterpartyVersions}
+    version = pickVersion(counterpartyVersions)
+    connection = ConnectionEnd{state, counterpartyConnectionIdentifier, counterpartyPrefix,
+                               clientIdentifier, counterpartyClientIdentifier, version}
+    abortTransactionUnless(connection.verifyConnectionState(proofHeight, proofInit, counterpartyConnectionIdentifier, expected))
+    abortTransactionUnless(connection.verifyClientConsensusState(
+      proofHeight, proofConsensus, counterpartyClientIdentifier, consensusHeight, expectedConsensusState))
+    previous = provableStore.get(connectionPath(desiredIdentifier))
+    abortTransactionUnless(
+      (previous === null) ||
+      (previous.state === INIT &&
+        previous.counterpartyConnectionIdentifier === counterpartyConnectionIdentifier &&
+        previous.counterpartyPrefix === counterpartyPrefix &&
+        previous.clientIdentifier === clientIdentifier &&
+        previous.counterpartyClientIdentifier === counterpartyClientIdentifier &&
+        previous.version === version))
+    identifier = desiredIdentifier
+    state = TRYOPEN
+    provableStore.set(connectionPath(identifier), connection)
+    addConnectionToClient(clientIdentifier, identifier)
+}
+```
+
+
+### 2. Using abstractions
+
+We will use the following three abstractions to parametrize the `connOpenTry` handler:
+
+- The `Party` type:
+
+```typescript
+interface Party {
+  connectionIdentifier: Identifier
+  clientIdentifier: Identifier
+  prefix: CommitmentPrefix
+}
+```
+
+- A `Connection` type:
+
+```typescript
+interface Connection {
+  state: ConnectionState
+  localParty: Party
+  remoteParty: Party
+  version: string | []string
+}
+```
+
+- and a `Proof` type:
+
+```typescript
+interface Proof {
+  content: CommitmentProof,
+  height: uint64,
+}
+```
+
+##### New Handler (version 2)
+
+Now we redefine `connOpenTry` handler as follows:
+
+```typescript
+function connOpenTry(
+  e: Connection,
+  proofInit: Proof,
+  proofConsensus: Proof) {
```

Comment:
I think that you are missing consensusHeight. ProofConsensus is a proof for that piece of data. Maybe it's better calling Connection ConnectionEnd.

Reply:
I defined the `Proof` type (see on line 82 just above your comment) to include the height. Including that definition here:

```
interface Proof {
   content: CommitmentProof,
   height: uint64,
 }
```